### PR TITLE
CPP SYCL interface deprecation

### PR DESCRIPTION
**Pull request by ethanglaser**
# Description

- break code when SyclExecutionContextImpl is used from oneDAL, but flag to maintain python use
- removed all cpp sycl usage in oneDAL oneapi interfaces knn, kmeans, linreg (via execution guard context removal) to resolve failures
- removed cpp sycl examples entirely and updated makefile accordingly
- removed samples/daal/cpp/oneccl

### CODE REVIEW

**Comment by ethanglaser** [comment]
Latest job: www.intel-ci.intel.com/ee468319-5a91-f191-85d9-a4bf010d0e2e

File: cpp/daal/include/services/internal/sycl/execution_context_sycl.h [hunk 1]
```diff
@@ -195,12 +195,13 @@ class OpenClKernelFactory : public Base, public ClKernelFactoryIface
...
 };
 
-class SyclExecutionContextImpl : public Base, public ExecutionContextIface
+class [[deprecated]] SyclExecutionContextImpl : public Base, public ExecutionContextIface
 {
 public:
     explicit SyclExecutionContextImpl(const ::sycl::queue & deviceQueue)
         : _deviceQueue(deviceQueue), _kernelFactory(_deviceQueue), _kernelScheduler(_deviceQueue)
     {
+        throw std::runtime_error("CPP SYCL interfaces have been deprecated as of 2024.0 release.");
```
Comment (samir-nasibli) [inline cpp/daal/include/services/internal/sycl/execution_context_sycl.h:204]: Please, remove runtime errors. Just leave deprecation messages

**Review by samir-nasibli — CHANGES_REQUESTED** [review]

File: cpp/daal/include/services/internal/sycl/execution_context_sycl.h [hunk 2]
```diff
@@ -195,12 +195,17 @@ class OpenClKernelFactory : public Base, public ClKernelFactoryIface
     ::sycl::queue & _deviceQueue;
 };
 
-class SyclExecutionContextImpl : public Base, public ExecutionContextIface
+class [[deprecated("CPP SYCL interfaces have been deprecated as of 2024.0 release.")]] SyclExecutionContextImpl : public Base,
+                                                                                                                  public ExecutionContextIface
 {
 public:
-    explicit SyclExecutionContextImpl(const ::sycl::queue & deviceQueue)
+    explicit SyclExecutionContextImpl(const ::sycl::queue & deviceQueue, const bool fromPython = false)
```
Comment (samir-nasibli) [inline cpp/daal/include/services/internal/sycl/execution_context_sycl.h:202]: Why we are doing this in runtime? 
This is not required.
You already added deprecation warnings.

**Review by samir-nasibli — DISMISSED** [review]

File: cpp/daal/include/services/internal/sycl/execution_context_sycl.h (hunk 2 above)
Comment (ethanglaser) [reply cpp/daal/include/services/internal/sycl/execution_context_sycl.h:202]: This approach was approved by managers during sync

**Comment by ethanglaser** [comment]
More relevant job: http://intel-ci.intel.com/ee4d9115-df5b-f15b-80a4-a4bf010d0e2e

File: cpp/daal/include/services/internal/sycl/execution_context_sycl.h [hunk 3]
```diff
@@ -195,12 +195,17 @@ class OpenClKernelFactory : public Base, public ClKernelFactoryIface
...
         : _deviceQueue(deviceQueue), _kernelFactory(_deviceQueue), _kernelScheduler(_deviceQueue)
     {
+        if (!fromPython)
+        {
+            throw std::runtime_error("CPP SYCL interfaces have been deprecated as of 2024.0 release.");
```
Comment (napetrov) [inline cpp/daal/include/services/internal/sycl/execution_context_sycl.h:207]:
```suggestion
            throw std::runtime_error("CPP SYCL interfaces have been removed as of 2024.0 release.");
```

**Comment by ethanglaser** [comment]
GPU CI with all 3 branches: http://intel-ci.intel.com/ee50d084-17b5-f1ed-808f-a4bf010d0e2e

**Review by KulikovNikita — APPROVED** [review]
Looks fine to me

File: cpp/daal/include/services/internal/execution_context.h [hunk 4]
```diff
@@ -127,10 +127,12 @@ class SyclExecutionContext : public ExecutionContext
...
      *  \param[in] deviceQueue SYCL* queue object to the device that is selected to perform computations
      */
-    SyclExecutionContext(const ::sycl::queue & deviceQueue) : ExecutionContext(createContext(deviceQueue), !deviceQueue.get_device().is_cpu()) {}
+    SyclExecutionContext(const ::sycl::queue & deviceQueue, const bool fromPython = false)
+        : ExecutionContext(createContext(deviceQueue, fromPython), !deviceQueue.get_device().is_cpu())
+    {}
 
 private:
-    static daal::services::internal::sycl::ExecutionContextIface * createContext(const ::sycl::queue & queue)
+    static daal::services::internal::sycl::ExecutionContextIface * createContext(const ::sycl::queue & queue, const bool fromPython = false)
```
Comment (samir-nasibli) [inline cpp/daal/include/services/internal/execution_context.h:135]: We already deprecated this function, I really don't see any reason for such update.
We can leave the signature as is, no updates its API required.

**Review by samir-nasibli — APPROVED** [review]

**Pull request merged**